File: ts-go-runtypes/internal/compiler/resolver/pure_fn_report_test.go
package resolver_test

import (
	"encoding/json"
	"os"
	"path/filepath"
	"testing"

	"github.com/mionkit/mion/ts-go-runtypes/internal/compiler/program"
	"github.com/mionkit/mion/ts-go-runtypes/internal/compiler/resolver"
	"github.com/mionkit/mion/ts-go-runtypes/internal/protocol"
)

// reportSources is the shared program for the report integration tests: both
// pure-fn lanes and both forms, so the emitted report covers the whole matrix.
var reportSources = map[string]string{
	"runtypes.d.ts": anonPureFnDTS,
	"a.ts": `import {registerPureFnFactory, registerAnonymousPureFn} from '@mionjs/run-types';
export const nf = registerPureFnFactory('acme::mul', (utl) => function _mul(x: number, y: number) { return x * y; });
export const ad = registerAnonymousPureFn((n: number): number => n * 2);
`,
}

// setupReport builds a report-enabled inline resolver (PureFnReportWire +
// PureFnReportFile), matching how the plugin forwards `pureFnReport: true`.
func setupReport(t *testing.T, moduleMode string) *resolver.Session {
	t.Helper()
	return setupReportIn(t, moduleMode, t.TempDir())
}

// setupReportIn is setupReport pinned to a caller-chosen output root, for the
// tests that read the generated report back off disk. The root is session
// config (the spawn-time --gen-dir), not a per-request field.
func setupReportIn(t *testing.T, moduleMode, genDir string) *resolver.Session {
	t.Helper()
	return setupInlineWith(t, reportSources, func(programOpts *program.Options, resolverOpts *resolver.Options) {
		programOpts.SingleThreaded = true
		resolverOpts.SingleThreaded = true
		resolverOpts.PureFnReportWire = true
		resolverOpts.PureFnReportFile = true
		resolverOpts.ModuleMode = moduleMode
		resolverOpts.GenDir = genDir
	})
}

// TestPureFnReport_GenerateWritesJsonAndResponse verifies OpGenerate populates
// Response.PureFnSites AND writes the JSON report file, that the file parses
// back to the same records, and that the report file is NOT part of the module
// manifest (it is data, not a module).
func TestPureFnReport_GenerateWritesJsonAndResponse(t *testing.T) {
	outDir := t.TempDir()
	r := setupReportIn(t, "", outDir)
	gen := r.Dispatch(protocol.Request{Op: protocol.OpGenerate})
	if gen.Error != "" {
		t.Fatalf("generate: %s", gen.Error)
	}
	if len(gen.PureFnSites) != 2 {
		t.Fatalf("expected 2 report records on the response, got %d: %+v", len(gen.PureFnSites), gen.PureFnSites)
	}

	// The report lives INSIDE types/, alongside the generated cache modules, so
	// it inherits that dir's .gitignore like every other regenerated artifact.
	reportPath := filepath.Join(outDir, "types", "pure-fns-report.json")
	raw, err := os.ReadFile(reportPath)
	if err != nil {
		t.Fatalf("report file not written: %v", err)
	}
	var fromDisk []protocol.PureFnSite
	if err := json.Unmarshal(raw, &fromDisk); err != nil {
		t.Fatalf("report file is not valid JSON: %v", err)
	}
	if len(fromDisk) != len(gen.PureFnSites) {
		t.Fatalf("disk report has %d records, response has %d", len(fromDisk), len(gen.PureFnSites))
	}

	// Keys on disk match the response keys AND the injected registry ids.
	diskKeys := map[string]bool{}
	for _, site := range fromDisk {
		diskKeys[site.Key] = true
		if site.CalleeModule != "@mionjs/run-types" {
			t.Errorf("record %s calleeModule = %q, want @mionjs/run-types", site.Key, site.CalleeModule)
		}
	}
	for _, site := range gen.PureFnSites {
		if !diskKeys[site.Key] {
			t.Errorf("response key %s missing from disk report", site.Key)
		}
	}
	// Named + anonymous both present.
	if !diskKeys["acme::mul"] {
		t.Errorf("named entry acme::mul missing from report keys %v", diskKeys)
	}

	// The report file is DATA, not a generated module — even though it sits
	// inside types/, it never enters the module manifest and is never a `.js`
	// module (so never resolvable as an rtmod:/ specifier nor GC'd by the
	// stale-module prune, which only touches *.js).
	for _, basename := range gen.Generated {
		if basename == "pure-fns-report" || filepath.Base(basename) == "pure-fns-report.json" {
			t.Fatalf("report file leaked into the module manifest: %q", basename)
		}
	}

	// A second generate must succeed (the report file inside types/ is never
	// inspected by the output-dir guard, and the stale-module prune skips it)
	// and rewrite the report.
	gen2 := r.Dispatch(protocol.Request{Op: protocol.OpGenerate})
	if gen2.Error != "" {
		t.Fatalf("second generate must succeed with the report file present, got: %s", gen2.Error)
	}
	if _, err := os.Stat(reportPath); err != nil {
		t.Fatalf("report file must survive a second generate (stale-module prune must skip it): %v", err)
	}
}

// TestPureFnReport_LayoutIndependent verifies the report shape (keys) is
// identical across moduleMode while the per-record `module` field carries the
// actual layout: per-entry pf/<ns>/<fn> vs the single `pf` bundle.
func TestPureFnReport_LayoutIndependent(t *testing.T) {
	perEntry := setupReport(t, "").Dispatch(protocol.Request{Op: protocol.OpGenerate})
	bundled := setupReport(t, "allSingle").Dispatch(protocol.Request{Op: protocol.OpGenerate})
	if perEntry.Error != "" || bundled.Error != "" {
		t.Fatalf("generate errors: %q / %q", perEntry.Error, bundled.Error)
	}
	if len(perEntry.PureFnSites) != len(bundled.PureFnSites) {
		t.Fatalf("record count differs across module modes: %d vs %d", len(perEntry.PureFnSites), len(bundled.PureFnSites))
	}
	keyModuleBundled := map[string]string{}
	for _, site := range bundled.PureFnSites {
		keyModuleBundled[site.Key] = site.Module
		if site.Module != "pf" {
			t.Errorf("allSingle: %s module = %q, want pf bundle", site.Key, site.Module)
		}
	}
	for _, site := range perEntry.PureFnSites {
		if _, ok := keyModuleBundled[site.Key]; !ok {
			t.Errorf("key %s present per-entry but missing in allSingle report", site.Key)
		}
		if site.Module == "" || site.Module == "pf" {
			t.Errorf("default mode: %s module = %q, want per-entry pf/<ns>/<fn>", site.Key, site.Module)
		}
	}
}

// TestPureFnReport_ScanDelta verifies OpScanFiles carries the report delta for
// the rescanned files (the plugin's update-lane callback source).
func TestPureFnReport_ScanDelta(t *testing.T) {
	r := setupReport(t, "")
	scan := r.Dispatch(protocol.Request{Op: protocol.OpScanFiles, Files: []string{"a.ts"}})
	if scan.Error != "" {
		t.Fatalf("scanFiles: %s", scan.Error)
	}
	if len(scan.PureFnSites) != 2 {
		t.Fatalf("scan delta expected 2 records for a.ts, got %d: %+v", len(scan.PureFnSites), scan.PureFnSites)
	}
}

// TestPureFnReport_OffByDefault verifies zero cost when the option is off: no
// PureFnSites on the response and no report file written.
func TestPureFnReport_OffByDefault(t *testing.T) {
	// Pin the output root so the "no report file" assertion checks the dir this
	// session actually writes to, rather than passing vacuously.
	outDir := t.TempDir()
	r := setupInlineWith(t, reportSources, func(programOpts *program.Options, resolverOpts *resolver.Options) {
		programOpts.SingleThreaded = true
		resolverOpts.SingleThreaded = true
		resolverOpts.GenDir = outDir
	})
	gen := r.Dispatch(protocol.Request{Op: protocol.OpGenerate})
	if gen.Error != "" {
		t.Fatalf("generate: %s", gen.Error)
	}
	if len(gen.PureFnSites) != 0 {
		t.Fatalf("report disabled but response carried %d records", len(gen.PureFnSites))
	}
	if _, err := os.Stat(filepath.Join(outDir, "types", "pure-fns-report.json")); err == nil {
		t.Fatalf("report disabled but file was written")
	}
}

File: packages/devtools/test/pure-fn-report.test.ts
// Pure-fn build report — the structured, layout-independent record of every
// pure fn a build generated, for host tooling that relocates pure-fn bodies
// across bundles (mion's cross-bundle batch transport). This suite
// drives the SHARED unplugin factory through the Rollup adapter (proving the
// callback is universal, NOT a vite-only hook) against a self-contained
// on-disk fixture program, and asserts:
//   - onPureFnReport fires once after buildStart with phase 'build', covering
//     both lanes (named `<ns>::name` + anonymous `rt::<hash>`) and both forms.
//   - the JSON report file round-trips (write → parse → keys match) under BOTH
//     moduleMode 'default' (per-entry pf modules) and 'allSingle' (one pf
//     bundle), with an identical report shape and the correct `module` field.
//   - the report file never collides with a generated module nor appears under
//     types/ (it is data, not a module).
import {afterEach, beforeEach, describe, expect, it} from 'vitest';
import path from 'node:path';
import os from 'node:os';
import fs from 'node:fs';
import runtypesRollup from '../src/runtypes/rollup.ts';
import type {PureFnSite} from '../src/core/protocol.ts';
import {BIN, hasBinary, writeMarkerPackage} from './helpers/inline.ts';

let FIXTURE_DIR = '';

const TSCONFIG = JSON.stringify({
  compilerOptions: {
    target: 'ES2022',
    module: 'ESNext',
    moduleResolution: 'bundler',
    strict: true,
    skipLibCheck: true,
    types: [],
  },
  include: ['*.ts'],
});

// Both lanes, both forms: named+factory, named+direct, anonymous+direct.
const CONSUMER = `import {registerPureFnFactory, registerPureFn, registerAnonymousPureFn} from '@mionjs/run-types';
export const nf = registerPureFnFactory('rep::mul', (utl) => function _mul(x: number, y: number) { return x * y; });
export const nd = registerPureFn('rep::neg', function _neg(x: number) { return -x; });
export const ad = registerAnonymousPureFn(function _double(n: number): number { return n * 2; });
`;

const ctx = {
  error(message: string): never {
    throw new Error(message);
  },
  warn(): void {},
};

const callHook = (hook: any, thisArg: unknown, ...args: unknown[]): unknown =>
  typeof hook === 'function' ? hook.apply(thisArg, args) : hook.handler.apply(thisArg, args);

function makePlugin(extra: Record<string, unknown>) {
  return runtypesRollup({
    binary: BIN,
    cwd: FIXTURE_DIR,
    tsconfig: 'tsconfig.json',
    genDir: path.join(FIXTURE_DIR, '.mion'),
    ...extra,
  }) as any;
}

describe('pure-fn build report', () => {
  const register = hasBinary() ? it : it.skip;

  beforeEach(() => {
    FIXTURE_DIR = fs.mkdtempSync(path.join(os.tmpdir(), 'rt-pf-report-'));
    fs.writeFileSync(path.join(FIXTURE_DIR, 'tsconfig.json'), TSCONFIG);
    writeMarkerPackage(FIXTURE_DIR);
    fs.writeFileSync(path.join(FIXTURE_DIR, 'consumer.ts'), CONSUMER);
  });
  afterEach(() => fs.rmSync(FIXTURE_DIR, {recursive: true, force: true}));

  register('onPureFnReport fires on the rollup adapter with both lanes + forms; JSON file round-trips', async () => {
    const calls: Array<{phase: string; sites: PureFnSite[]}> = [];
    const plugin = makePlugin({
      pureFnReport: 'file',
      onPureFnReport: (sites: PureFnSite[], phase: 'build' | 'update') => calls.push({phase, sites}),
    });
    try {
      await callHook(plugin.buildStart, ctx);
    } finally {
      try {
        await callHook(plugin.buildEnd, ctx);
      } catch {
        // best-effort
      }
    }

    // The callback fired once, with the whole-program report.
    expect(calls.length, 'onPureFnReport must fire on buildStart (universal hook)').toBe(1);
    expect(calls[0].phase).toBe('build');
    const sites = calls[0].sites;
    const byKey = new Map(sites.map((s) => [s.key, s]));

    // Named lane keys present verbatim.
    expect(byKey.has('rep::mul'), `named factory rep::mul missing: ${[...byKey.keys()]}`).toBe(true);
    expect(byKey.has('rep::neg'), 'named direct rep::neg missing').toBe(true);
    // Anonymous lane: exactly one rt::<hash> record.
    const anon = sites.filter((s) => s.lane === 'anonymous');
    expect(anon.length, 'one anonymous record').toBe(1);
    expect(anon[0].key).toMatch(/^rt::[A-Za-z0-9_-]+$/);

    // Forms + callee attribution (primitive registrar → @mionjs/run-types).
    expect(byKey.get('rep::mul')!.form).toBe('factory');
    expect(byKey.get('rep::mul')!.lane).toBe('named');
    expect(byKey.get('rep::mul')!.calleeName).toBe('registerPureFnFactory');
    expect(byKey.get('rep::mul')!.calleeModule).toBe('@mionjs/run-types');
    expect(byKey.get('rep::neg')!.form).toBe('direct');
    expect(anon[0].form).toBe('direct');

    // The self-contained payload rides inline (code + paramNames) — no need to
    // read generated modules. Default emitMode ships the code body string.
    expect(byKey.get('rep::mul')!.paramNames).toEqual(['utl']);
    expect(byKey.get('rep::mul')!.code, 'factory code should be present in default emit mode').toBeTruthy();

    // JSON file round-trips: write → parse → keys match the injected report.
    // The report lives INSIDE types/, alongside the generated cache modules, so
    // it inherits that dir's .gitignore (`*`) exactly like every cache module.
    const typesDir = path.join(FIXTURE_DIR, '.mion', 'types');
    const reportPath = path.join(typesDir, 'pure-fns-report.json');
    expect(fs.existsSync(reportPath), 'pure-fns-report.json must be written under types/ on generate').toBe(true);
    const fromDisk = JSON.parse(fs.readFileSync(reportPath, 'utf8')) as PureFnSite[];
    expect(new Set(fromDisk.map((s) => s.key))).toEqual(new Set(sites.map((s) => s.key)));

    // It is data, not a module: no `.js` module collides with it, and the
    // generated `types/.gitignore` (`*`) covers it just like the cache modules.
    const typeFiles = fs.readdirSync(typesDir);
    expect(typeFiles).not.toContain('pure-fns-report.js');
    expect(fs.readFileSync(path.join(typesDir, '.gitignore'), 'utf8')).toContain('*');
  });

  register('report shape is identical across moduleMode; module field carries the layout', async () => {
    async function reportFor(moduleMode: string): Promise<PureFnSite[]> {
      let captured: PureFnSite[] = [];
      const plugin = makePlugin({
        moduleMode,
        pureFnReport: 'file',
        onPureFnReport: (sites: PureFnSite[]) => (captured = sites),
      });
      try {
        await callHook(plugin.buildStart, ctx);
      } finally {
        try {
          await callHook(plugin.buildEnd, ctx);
        } catch {
          // best-effort
        }
      }
      return captured;
    }

    const perEntry = await reportFor('default');
    const bundled = await reportFor('allSingle');

    // Same keys either way — the report shape does not depend on moduleMode.
    expect(new Set(bundled.map((s) => s.key))).toEqual(new Set(perEntry.map((s) => s.key)));

    // default: per-entry pf/<ns>/<fn>. allSingle: the single `pf` bundle.
    for (const s of perEntry) {
      expect(s.module, `${s.key} per-entry module`).toMatch(/^pf\//);
    }
    for (const s of bundled) {
      expect(s.module, `${s.key} allSingle module`).toBe('pf');
    }
  });

  register('report data flows to the callback even with no JSON file (callback-only)', async () => {
    let captured: PureFnSite[] = [];
    const plugin = makePlugin({
      // No `pureFnReport` file option — only the in-process callback.
      onPureFnReport: (sites: PureFnSite[]) => (captured = sites),
    });
    try {
      await callHook(plugin.buildStart, ctx);
    } finally {
      try {
        await callHook(plugin.buildEnd, ctx);
      } catch {
        // best-effort
      }
    }
    expect(captured.length, 'callback receives records without a file being requested').toBe(3);
    // ...and no JSON file was written (data-only).
    expect(fs.existsSync(path.join(FIXTURE_DIR, '.mion', 'types', 'pure-fns-report.json'))).toBe(false);
  });

  // Validation runs at plugin construction (no binary needed), so this stays a
  // plain `it` — an unknown tri-state value must fail loudly, not silently.
  it('rejects an unknown pureFnReport value', () => {
    expect(() => makePlugin({pureFnReport: 'bogus'})).toThrow(/pureFnReport/);
  });
});
